### ISSUE TITLE
[Performance] The functions `core::ptr::copy` and `core::ptr::copy_nonoverlapping` for copying memory are slow

### ISSUE THREAD

**Issue by liqinggd** [asterinas/asterinas#839]
I conducted benchmark tests on these functions to verify their performance. The experimental results show that their latency is about 169 CPU cycles when copying 4KB of memory. Meanwhile, the latency of memory copy implemented based on the `rep movsb` instruction is only about 67 CPU cycles.

Here is the sample code:
```rust
pub unsafe fn read_tsc() -> u64 {
    let low: u32;
    let high: u32;
    core::arch::asm!(
        "rdtsc",
        out("eax") low,
        out("edx") high,
        options(nomem, nostack, preserves_flags),
    );
    ((high as u64) << 32) | (low as u64)
}

pub unsafe fn fast_copy(src: *const u8, dst: *mut u8, count: usize) {
    core::arch::asm!(
        "push rcx",
        "push rsi",
        "push rdi",
        "rep movsb",
        "pop rdi",
        "pop rsi",
        "pop rcx",
        in("rcx") count,
        in("rsi") src,
        in("rdi") dst,
    );
}

    #[ktest]
    fn memcpy_benchmark() {
        const SIZE: usize = 4096;
        const NUM: u64 = 100_0000;

        let src = vec![2u8; SIZE];
        let mut dst = vec![1u8; SIZE];
        
        black_box(unsafe { fast_copy(src.as_ptr(), dst.as_mut_ptr(), SIZE) });
        assert_eq!(src, dst);
        

        let start = unsafe { read_tsc() };
        for _ in 0..NUM {
            black_box(unsafe { core::ptr::copy(src.as_ptr(), dst.as_mut_ptr(), SIZE) });
        }
        let end = unsafe { read_tsc() };
        let avg = (end - start) / NUM;
        log::error!("memcpy with ptr::copy avg latency is: {}", avg);

        let start = unsafe { read_tsc() };
        for _ in 0..NUM {
            black_box(unsafe { fast_copy(src.as_ptr(), dst.as_mut_ptr(), SIZE) });
        }
        let end = unsafe { read_tsc() };
        let avg = (end - start) / NUM;
        log::error!("memcpy with fast copy avg latency is: {}", avg);
    }
```

**Comment by StevenJiang1110** [comment]
I also conducted a test for the `fast_copy` function (the version provided in this issue, not in PR #845). I was curious about the relationship between the execution time and the buffer size. Here are the results:

The red points represent the results of `ptr::copy_overlapping`, the green points represent the results of `ptr::copy`, and the blue points represent the results of `fast_copy`. As the results of `ptr::copy_overlapping` and `ptr::copy` are nearly identical, the red and green points overlap. The X-axis represents the size of the buffer, while the Y-axis represents the time taken for the copy operation.

**BUFFER SIZE FROM 16 to 8192, STEP_SIZE=16**:

<img width="1147" alt="截屏2024-05-17 15 55 21" src="https://github.com/asterinas/asterinas/assets/27764680/bb2c4464-b655-4d33-9b6d-b3b99d977e82">

The results are rather surprising. There is an abnormal spike when the buffer size is around 4096 for all three versions.

For buffer sizes smaller than approximately 2000, the `ptr::copy` function performs faster than `fast_copy`. 

As the buffer size approaches 4096 but does not exceed it, `fast_copy` becomes faster than `ptr::copy`. 

For buffer sizes exceeding 4096, `fast_copy` and `ptr::copy` show similar performance.

Here is a closer look at the spike around 4096. The time taken for the copy operation is closely correlated with the buffer size.

**BUFFER SIZE FROM 3036 to 4256, STEP_SIZE=8**:

<img width="1130" alt="截屏2024-05-17 16 13 27" src="https://github.com/asterinas/asterinas/assets/27764680/df8b731a-cc50-4bd7-839e-a1ec5c236478">

**Comment by StevenJiang1110** [comment]
FYI, below is the code for conducting the test. Build with relase mode, LTO enabled.

```rust
use core::hint::black_box;
use std::time::Instant;

use plotlib::{
    page::Page,
    repr::Plot,
    style::{PointMarker, PointStyle},
    view::ContinuousView,
};

pub unsafe fn fast_copy(src: *const u8, dst: *mut u8, count: usize) {
    core::arch::asm!(
        "push rcx",
        "push rsi",
        "push rdi",
        "rep movsb",
        "pop rdi",
        "pop rsi",
        "pop rcx",
        in("rcx") count,
        in("rsi") src,
        in("rdi") dst,
    );
}

const NUM: u64 = 100_0000;

const INIT_SIZE: usize = 16;
const MAX_SIZE: usize = 8192;
const STEP_SIZE: usize = 16;

fn main() {
    let mut res_nonoverlap = Vec::with_capacity(MAX_SIZE / STEP_SIZE);
    let mut res_copy = Vec::with_capacity(MAX_SIZE / STEP_SIZE);
    let mut res_fast_copy = Vec::with_capacity(MAX_SIZE / STEP_SIZE);

    for size in (INIT_SIZE..MAX_SIZE).step_by(STEP_SIZE) {
        let src = vec![127u8; size];
        let mut dst = vec![254u8; size];

        black_box(unsafe { core::ptr::copy_nonoverlapping(src.as_ptr(), dst.as_mut_ptr(), size) });

        // test ptr::copy_nonoverlapping

        let now = Instant::now();
        for _ in 0..NUM {
            black_box(unsafe {
                core::ptr::copy_nonoverlapping(src.as_ptr(), dst.as_mut_ptr(), size)
            });
        }
        let duration1 = now.elapsed();
        res_nonoverlap.push((size, duration1.as_nanos()));

        // test ptr::copy

        let now = Instant::now();
        for _ in 0..NUM {
            black_box(unsafe { core::ptr::copy(src.as_ptr(), dst.as_mut_ptr(), size) });
        }
        let duration2 = now.elapsed();
        res_copy.push((size, duration2.as_nanos()));

        // test fast_copy

        let now = Instant::now();
        for _ in 0..NUM {
            black_box(unsafe { fast_copy(src.as_ptr(), dst.as_mut_ptr(), size) });
        }
        let duration3 = now.elapsed();
        res_fast_copy.push((size, duration3.as_nanos()));

        println!(
            "size = {}, nonoverlap = {:?}, ptr::copy = {:?}, fast copy = {:?}",
            size, duration1, duration2, duration3
        );
    }

    draw_slot(res_nonoverlap, res_copy, res_fast_copy)
}

fn draw_slot(
    res_nonoverlap: Vec<(usize, u128)>,
    res_copy: Vec<(usize, u128)>,
    res_fast_copy: Vec<(usize, u128)>,
) {
    let y_min1 = res_nonoverlap.iter().map(|(_, y0)| *y0).min().unwrap();
    let y_max1 = res_nonoverlap.iter().map(|(_, y0)| *y0).max().unwrap();

    let y_min2 = res_copy.iter().map(|(_, y0)| *y0).min().unwrap();
    let y_max2 = res_copy.iter().map(|(_, y0)| *y0).max().unwrap();

    let y_min3 = res_fast_copy.iter().map(|(_, y0)| *y0).min().unwrap();
    let y_max3 = res_fast_copy.iter().map(|(_, y0)| *y0).max().unwrap();

    let y_min = y_min1.min(y_min2).min(y_min3);
    let y_max = y_max1.max(y_max2).max(y_max3);

    let res_nonoverlap = res_nonoverlap
        .iter()
        .map(|(x0, y0)| (*x0 as f64, *y0 as f64))
        .collect();
    let s1: Plot = Plot::new(res_nonoverlap).point_style(
        PointStyle::new()
            .marker(PointMarker::Square) // setting the marker to be a square
            .colour("#FF0000")
            .size(2.0), // red
    );

    let res_copy = res_copy
        .iter()
        .map(|(x0, y0)| (*x0 as f64, *y0 as f64))
        .collect();
    let s2: Plot = Plot::new(res_copy).point_style(
        PointStyle::new()
            .marker(PointMarker::Circle) // setting the marker to be a square
            .colour("#00FF00")
            .size(2.0), // green
    );

    let res_fast_copy = res_fast_copy
        .iter()
        .map(|(x0, y0)| (*x0 as f64, *y0 as f64))
        .collect();
    let s3: Plot = Plot::new(res_fast_copy).point_style(
        PointStyle::new()
            .marker(PointMarker::Circle) // setting the marker to be a square
            .colour("#0000FF")
            .size(2.0), // blue
    );

    let v = ContinuousView::new()
        .add(s1)
        .add(s2)
        .add(s3)
        .x_range((INIT_SIZE - STEP_SIZE) as f64, (MAX_SIZE + 1) as f64)
        .y_range((y_min - 1) as f64, (y_max + 1) as f64)
        .x_label("the size of vector")
        .y_label("the copy time(ns)");

    Page::single(&v).save("scatter.svg").unwrap()
}

```

**Comment by yingdi-shan** [comment]
I suspect that L1 cache bank conflicts are the primary cause of this phenomenon. Modern CPUs utilize set-associative caches, where the placement of cache lines is determined by specific bits of the address. If our buffer address is aligned to a power of 2 and the buffer size approaches 4KB, more data may contain conflicting cache lines in the L1 cache, thereby impacting performance.

**Comment by lrh2000** [comment]
@StevenJiang1110 In your test code, `ptr::copy`/`ptr::copy_nonoverlapping` directly calls glibc's `memcpy`/`memmove` implementations. But these implementations are very complex (they even use some SSE/AVX instructions, see [here](https://elixir.bootlin.com/glibc/glibc-2.39/source/sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S#L889)). They are not feasible to be used in the kernel. So comparing them to a simple `rep; movsb` is not fair.

In a `no_std` environment, LLVM inserts built-in implementations for `memcpy`/`memmove` directly. The built-in implementations are simple enough to be used in the kernel. For example, `memcpy` is implemented as follows:
```gas
<_ZN17compiler_builtins3mem6memcpy17h31453540a835f19dE>:
mov    %rdi,%rax
mov    %eax,%r8d
neg    %r8d
and    $0x7,%r8d
cmp    %rdx,%r8
cmovae %rdx,%r8
mov    %r8,%rcx
rep movsb %ds:(%rsi),%es:(%rdi)      # align %rdi to the 8-byte boundary
sub    %r8,%rdx
mov    %rdx,%rcx
shr    $0x3,%rcx
rep movsq %ds:(%rsi),%es:(%rdi)      # copy quadwords
and    $0x7,%edx
mov    %rdx,%rcx
rep movsb %ds:(%rsi),%es:(%rdi)      # copy remaining bytes
ret
```

I still don't understand how this simple and straightforward implementation can ever be significantly slower than `rep; movsb` (as pointed by this issue), though. So from the assembly level, I don't get the point why `fast_copy_nonoverlapping` is better than `ptr::copy_nonoverlapping`, either (https://github.com/asterinas/asterinas/pull/845#issuecomment-2133751334).

**Comment by junyang-zh** [comment]
I don't support any of your claims about the performance but I'd rather say that the current `fast_copy_nonoverlapping` is a maintenance disaster. Much more special attentions needed because of it when doing #836 and so for the RISC-V introduction. Use the built-in one is an intuitive option either.

**Comment by lrh2000** [comment]
> the current fast_copy_nonoverlapping is a maintenance disaster

Yes, I think we need either (reproducible) experimental results (e.g. number of CPU cycles) or theoretical analysis (e.g. by analyzing assembly instructions), if not both, to justify why we need `fast_copy_nonoverlapping`. Unfortunately, I do not see any at this time.

Even if the above condition is met, I still think that having `fast_copy_nonoverlapping` is only a temporary workaround. The real problem needs to be fixed in LLVM (`ptr::copy_nonoverlapping` uses LLVM's compiler builtins). I don't see any benefit in maintaining it ourselves.

**Comment by lrh2000** [comment]
> memcpy with ptr::copy avg latency is: 4122
> memcpy with fast copy avg latency is: 31613

@junyang-zh You need to enable KVM to do performance tests.

**Comment by junyang-zh** [comment]
> You need to enable KVM to do performance tests.

@lrh2000 Fixed. By disabling KVM the fast version is significantly faster however.

**Comment by lrh2000** [comment]
> By enabling KVM the fast version is significantly faster however.

OK, that's pretty interesting.

Could you please check the generated assembler instructions as well? I'm wondering if it's the difference in CPU or the assembler instructions that are causing the problem.

If you can reproduce the problem, I would also suggest that you check the performance counters (which can be collected by the `perf` tools in Linux, but you need to have a userspace version) to find out the reason why the `ptr::copy_nonoverlapping` is slow. I had planned to do this, but I cannot reproduce the problem anyway.

**Comment by junyang-zh** [comment]
> Could you please check the generated assembler instructions as well?

For `ptr::copy_nonoverlapping`, it is:

```
   0xffffffff88089550 <+224>:   mov    $0x1000,%edx
   0xffffffff88089555 <+229>:   mov    %rbx,%rdi
   0xffffffff88089558 <+232>:   mov    %r14,%rsi
   0xffffffff8808955b <+235>:   call   0xffffffff880a35e0 <compiler_builtins::mem::memmove::memmove>
   0xffffffff88089560 <+240>:   dec    %r12
   0xffffffff88089563 <+243>:   jne    0xffffffff88089550 <_ZN11aster_frame2vm4test16memcpy_benchmark17hd177f71ff1>

...

Dump of assembler code for function _ZN17compiler_builtins3mem7memmove17hdfd9a905d75bdcebE:
   0xffffffff880a31a0 <+0>:     mov    %rdi,%rax
   0xffffffff880a31a3 <+3>:     mov    %rdi,%rcx
   0xffffffff880a31a6 <+6>:     sub    %rsi,%rcx
   0xffffffff880a31a9 <+9>:     mov    %eax,%r9d
   0xffffffff880a31ac <+12>:    neg    %r9d
   0xffffffff880a31af <+15>:    and    $0x7,%r9d
   0xffffffff880a31b3 <+19>:    cmp    %rdx,%r9
   0xffffffff880a31b6 <+22>:    cmovae %rdx,%r9
   0xffffffff880a31ba <+26>:    mov    %rdx,%r10
   0xffffffff880a31bd <+29>:    sub    %r9,%r10
   0xffffffff880a31c0 <+32>:    mov    %r10,%r8
   0xffffffff880a31c3 <+35>:    shr    $0x3,%r8
   0xffffffff880a31c7 <+39>:    and    $0x7,%r10d
   0xffffffff880a31cb <+43>:    cmp    %rdx,%rcx
   0xffffffff880a31ce <+46>:    jae    0xffffffff880a3203 <_ZN17compiler_builtins3mem7memmove17hdfd9a905d75bdcebE+>
   0xffffffff880a31d0 <+48>:    lea    (%rax,%rdx,1),%rdi
   0xffffffff880a31d4 <+52>:    dec    %rdi
   0xffffffff880a31d7 <+55>:    add    %rdx,%rsi
   0xffffffff880a31da <+58>:    dec    %rsi
   0xffffffff880a31dd <+61>:    mov    %r10,%rcx
   0xffffffff880a31e0 <+64>:    std    
   0xffffffff880a31e1 <+65>:    rep movsb %ds:(%rsi),%es:(%rdi)
   0xffffffff880a31e3 <+67>:    sub    $0x7,%rsi
   0xffffffff880a31e7 <+71>:    sub    $0x7,%rdi
   0xffffffff880a31eb <+75>:    mov    %r8,%rcx
   0xffffffff880a31ee <+78>:    rep movsq %ds:(%rsi),%es:(%rdi)
   0xffffffff880a31f1 <+81>:    test   %r9d,%r9d
   0xffffffff880a31f4 <+84>:    add    $0x7,%rsi
   0xffffffff880a31f8 <+88>:    add    $0x7,%rdi
   0xffffffff880a31fc <+92>:    mov    %r9d,%ecx
   0xffffffff880a31ff <+95>:    rep movsb %ds:(%rsi),%es:(%rdi)
   0xffffffff880a3201 <+97>:    cld    
   0xffffffff880a3202 <+98>:    ret    
   0xffffffff880a3203 <+99>:    mov    %r9,%rcx
   0xffffffff880a3206 <+102>:   mov    %rax,%rdi
   0xffffffff880a3209 <+105>:   rep movsb %ds:(%rsi),%es:(%rdi)
   0xffffffff880a320b <+107>:   mov    %r8,%rcx
   0xffffffff880a320e <+110>:   rep movsq %ds:(%rsi),%es:(%rdi)
   0xffffffff880a3211 <+113>:   mov    %r10,%rcx
   0xffffffff880a3214 <+116>:   rep movsb %ds:(%rsi),%es:(%rdi)
   0xffffffff880a3216 <+118>:   ret    
End of assembler dump.
```

**Comment by lrh2000** [comment]
> For ptr::copy_nonoverlapping
> compiler_builtins::mem::memmove::memmove

Well, `ptr::copy_nonoverlapping` should use `memcpy`, and `ptr::copy` should use `memmove`.

But that does not matter. Modulo some surrounding preparation code, the only critical part is:
```
   0xffffffff880a3209 <+105>:   rep movsb %ds:(%rsi),%es:(%rdi)
   0xffffffff880a320b <+107>:   mov    %r8,%rcx
   0xffffffff880a320e <+110>:   rep movsq %ds:(%rsi),%es:(%rdi)
   0xffffffff880a3211 <+113>:   mov    %r10,%rcx
   0xffffffff880a3214 <+116>:   rep movsb %ds:(%rsi),%es:(%rdi)
```

So I think it ends up just using `rep; movsb` and `rep; movsq`, which I can in no way understand to be significantly slower than a trivial `rep; movsb`.

**Comment by lrh2000** [comment]
```rust
        core::arch::asm!(
            "push rcx",
            "push rsi",
            "push rdi",
            "mov rcx, 8",
            "rep movsb",
            in("rsi") src,
            in("rdi") dst,
        );
        core::arch::asm!(
            "rep movsq",
            in("rcx") count,    // <X>
        );
```

Doesn't the first assembly code break some invariants, e.g. the stack pointer and some register contents (`rcx`)? So it's UB and I'm afraid that accessing `count` at `<X>` might take a wrong value unless you have checked the final assembly code to make sure it's ok.

**Comment by junyang-zh** [comment]
> Doesn't the first assembly code break some invariants, ...

Oh, sorry. That surely copied less bytes than expected. Here is a robust one:

Result:

```
memcpy with ptr::copy avg latency is: 211
memcpy with ptr::copy_nonoverlapping avg latency is: 212
memcpy with aligned ptr::copy_nonoverlapping avg latency is: 209
memcpy with fast copy avg latency is: 126
memcpy with fast copy #1 avg latency is: 233
memcpy with fast copy #2 avg latency is: 127
```

Seems mixing `movsq` and `movsb` is the root problem. This behavior varies among different CPUs.

Source:

```rust
mod test {
    use core::hint::black_box;
    use alloc::vec;

    pub unsafe fn read_tsc() -> u64 {
        let low: u32;
        let high: u32;
        core::arch::asm!(
            "rdtsc",
            out("eax") low,
            out("edx") high,
            options(nomem, nostack, preserves_flags),
        );
        ((high as u64) << 32) | (low as u64)
    }
    
    pub unsafe fn fast_copy(src: *const u8, dst: *mut u8, count: usize) {
        core::arch::asm!(
            "push rcx",
            "push rsi",
            "push rdi",
            "rep movsb",
            "pop rdi",
            "pop rsi",
            "pop rcx",
            in("rcx") count,
            in("rsi") src,
            in("rdi") dst,
        );
    }
    
    pub unsafe fn fast_copy1(src: *const u8, dst: *mut u8, count: usize) {
        core::arch::asm!(
            "push rcx",
            "push rsi",
            "push rdi",
            "sub rcx, 16",
            "shr rcx, 3",
            "push rcx",
            "mov rcx, 8",
            "rep movsb",
            in("rsi") src,
            in("rdi") dst,
            in("rcx") count,
        );
        core::arch::asm!(
            "pop rcx",
            "rep movsq",
            "mov rcx, 8",
            "rep movsb",
            "pop rdi",
            "pop rsi",
            "pop rcx",
        );
    }
    
    pub unsafe fn fast_copy2(src: *const u8, dst: *mut u8, count: usize) {
        let count = count / 8;
        core::arch::asm!(
            "push rcx",
            "push rsi",
            "push rdi",
            "rep movsq",
            "pop rdi",
            "pop rsi",
            "pop rcx",
            in("rcx") count,
            in("rsi") src,
            in("rdi") dst,
        );
    }
    
    #[ktest]
    fn memcpy_benchmark() {
        const SIZE: usize = 4096;
        const NUM: u64 = 100_0000;

        let src = vec![2u8; SIZE];
        let mut dst = vec![1u8; SIZE];
        
        black_box(unsafe { fast_copy(src.as_ptr(), dst.as_mut_ptr(), SIZE) });
        assert_eq!(src, dst);
        dst.fill(1);

        let start = unsafe { read_tsc() };
        for _ in 0..NUM {
            black_box(unsafe { core::ptr::copy(src.as_ptr(), dst.as_mut_ptr(), SIZE) });
        }
        let end = unsafe { read_tsc() };
        let avg = (end - start) / NUM;
        assert_eq!(src, dst);
        dst.fill(1);
        crate::early_println!("memcpy with ptr::copy avg latency is: {}", avg);

        let start = unsafe { read_tsc() };
        for _ in 0..NUM {
            black_box(unsafe { core::ptr::copy_nonoverlapping(src.as_ptr(), dst.as_mut_ptr(), SIZE) });
        }
        let end = unsafe { read_tsc() };
        let avg = (end - start) / NUM;
        assert_eq!(src, dst);
        dst.fill(1);
        crate::early_println!("memcpy with ptr::copy_nonoverlapping avg latency is: {}", avg);
        
        let start = unsafe { read_tsc() };
        for _ in 0..NUM {
            black_box(unsafe { core::ptr::copy_nonoverlapping(src.as_ptr() as *const u64, dst.as_mut_ptr() as *mut u64, SIZE / 8) });
        }
        let end = unsafe { read_tsc() };
        let avg = (end - start) / NUM;
        assert_eq!(src, dst);
        dst.fill(1);
        crate::early_println!("memcpy with aligned ptr::copy_nonoverlapping avg latency is: {}", avg);

        let start = unsafe { read_tsc() };
        for _ in 0..NUM {
            black_box(unsafe { fast_copy(src.as_ptr(), dst.as_mut_ptr(), SIZE) });
        }
        let end = unsafe { read_tsc() };
        let avg = (end - start) / NUM;
        assert_eq!(src, dst);
        dst.fill(1);
        crate::early_println!("memcpy with fast copy avg latency is: {}", avg);
        
        let start = unsafe { read_tsc() };
        for _ in 0..NUM {
            black_box(unsafe { fast_copy1(src.as_ptr(), dst.as_mut_ptr(), SIZE) });
        }
        let end = unsafe { read_tsc() };
        let avg = (end - start) / NUM;
        assert_eq!(src, dst);
        dst.fill(1);
        crate::early_println!("memcpy with fast copy #1 avg latency is: {}", avg);

        
        let start = unsafe { read_tsc() };
        for _ in 0..NUM {
            black_box(unsafe { fast_copy2(src.as_ptr(), dst.as_mut_ptr(), SIZE) });
        }
        let end = unsafe { read_tsc() };
        let avg = (end - start) / NUM;
        assert_eq!(src, dst);
        dst.fill(1);
        crate::early_println!("memcpy with fast copy #2 avg latency is: {}", avg);
    }
}
```

**Comment by junyang-zh** [comment]
> So I think it ends up just using rep; movsb and rep; movsq, which I can in no way understand to be significantly slower than a trivial rep; movsb.

@lrh2000 Well there seems to be an interesting (yet proven) finding that our CPUs may suffer from more micro-code level state checking when combining `movsb` and `movsq` compared to using `movsb` alone.

**Comment by lrh2000** [comment]
@junyang-zh :+1: Thank you for your investigation on this issue. I have checked the Intel manual and I think your conclusion makes sense.

Intel has three important optimizations for `rep; movsb`:
 - The first one is "Enhanced REP MOVSB and STOSB", but it is available since the Ivy Bridge microarchitecture, which was released in 2012. So today we can assume that it always exists.
> Beginning with processors based on Ivy Bridge microarchitecture, REP string operation using MOVSB and
STOSB can provide both flexible and high-performance REP string operations for software in common
situations like memory copy and set operations.
 - The second one is "Fast Short REP MOVSB":
> Beginning with processors based on Ice Lake Client microarchitecture, REP MOVSB performance of short
operations is enhanced. The enhancement applies to string lengths between 1 and 128 bytes long.
 - The third one is "Fast Zero Length REP MOVSB":
> REP MOVSB performance of zero length operations is enhanced. The latency of a zero length REP MOVSB
is now the same as the latency of lengths 1 to 128 bytes. When both Fast Short REP MOVSB and Fast Zero
Length REP MOVSB features are enabled, REP MOVSB performance is flat 9 cycles per operation, for all
strings 0-128 byte long whose source and destination operands reside in the processor first level cache.

Unfortunately, without the the last two optimizations, `rep; movsb` can be very slow when the number of bytes is small:
> String operations - string operations can produce a large amount of microcode. In some cases
there are assists which can occur due to string operations such as REP MOVSB with trip count
greater than 3, which costs 70+ cycles.

According to [Intel's official website](https://www.intel.com/content/www/us/en/products/sku/199354/intel-xeon-gold-5220r-processor-35-75m-cache-2-20-ghz/specifications.html), the Intel® Xeon® Gold 5220R processor belongs to the 2nd Xeon generation and uses the Cascade Lake microarchitecture. It is before the 3rd Xeon generation and the Ice Lake microarchitecture, according to [Wikipedia](https://en.wikipedia.org/wiki/List_of_Intel_CPU_microarchitectures). Thus, it does not have the last two optimizations mentioned above, which means that the short-length and zero-length `rep; movsb` can be very slow, causing the problem in this issue.

I'm using a 13th Gen Intel(R) Core(TM) i5-13500H processor, which is newer and has the last two optimizations. So it is expected that I won't be able to reproduce this issue.

Reference: [Intel® 64 and IA-32 Architectures Optimization Reference Manual Volume 1](https://www.intel.com/content/www/us/en/content-details/671488/intel-64-and-ia-32-architectures-optimization-reference-manual-volume-1.html)

@junyang-zh You can additionally check whether `cat /proc/cpuinfo | grep 'fsrm'` finds anything in your server. It stands for the "Fast Short REP MOVSB" feature, so I think it won't succeed as the CPU on your server should not have this feature.

**Comment by junyang-zh** [comment]
> You can additionally check whether `cat /proc/cpuinfo | grep 'fsrm'` finds anything in your server.

No the "Fast Short REP MOVSB" feature is not introduced in our servers.

Thanks very much about your confirmation on the ISA side. I think there's enough evidence that the implementation of short `rep; movsb` is the root cause of out performance issue. As it is not the responsibility for Asterinas to fill the gap between LLVM and old hardwares, I suggest reverting #845.

@tatetian @StevenJiang1110 @liqinggd What do you think?

**Comment by lrh2000** [comment]
A bit more... Rust actually rewrites LLVM's compiler-rt as [rust-lang/compiler_builtins](https://github.com/rust-lang/compiler-builtins). That's where `memcpy`/`memmove` is implemented.

> As it is not the responsibility for Asterinas to fill the gap between LLVM and old hardwares, I suggest reverting https://github.com/asterinas/asterinas/pull/845.

I second this. My main reason is that we should fix [rust-lang/compiler_builtins](https://github.com/rust-lang/compiler-builtins) instead, as **this will benefit all implicit `ptr::copy` calls (e.g. `Vec::clone`, `slice::copy_from_slice`, and many, many more)**.

But sending the fix patch upstream takes some time, so leaving the current workaround here is fine with me (of course, reverting it immediately is fine with me too, I have no opinion on that).

I've made a draft commit: https://github.com/lrh2000/compiler-builtins/commit/ce7950cfb09a0051173cff9ffb250bb1ec3429b7

This commit shows performance improvement for copying 4KiB data even on my newer CPU:
Before: ``` test memcpy_rust_4096                 ... bench:          33.15 ns/iter (+/- 0.46) = 124121 MB/s ```
After: ``` test memcpy_rust_4096                 ... bench:          31.56 ns/iter (+/- 0.16) = 132129 MB/s ```

I need the performance improvement statistics on the old CPU and do more but similar work on related memory functions before sending it upstream.

**Comment by tatetian** [comment]
@lrh2000 To clarify, are you able to reproduce the problem or not? You have proposed to fix the problem with changes to compiler_builtins, but haven't yet confirmed @StevenJiang1110 's finding. So I am confused with the performance results.

> I second this. My main reason is that we should fix [rust-lang/compiler_builtins](https://github.com/rust-lang/compiler-builtins) instead, as **this will benefit all implicit `ptr::copy` calls (e.g. `Vec::clone`, `slice::copy_from_slice`, and many, many more)**.

Of course. Fixing the `compiler_builtin` is the best option in the long term. But it could take a very long time before Asterinas can use the optimized version of `compiler_builtin`. So at least, accepting `fast_copy` as a workaround is reasonable, right? Without optimized memcopy, how we are supposed to do benchmarks on some picky machines _now_?

**Comment by junyang-zh** [comment]
> Without optimized memcopy, how we are supposed to do benchmarks on some picky machines now?

#890

**Comment by junyang-zh** [comment]
> I need the performance improvement statistics on the old CPU and do more but similar work on related memory functions before sending it upstream.

I'll help you validate it on our machines.

**Comment by liqinggd** [comment]
Let me explain why we consider using "rep movsb" directly for memory copying. 

This is because, during our memory copy performance testing, our comparison target is the **Linux kernel**, and we have confirmed that it directly uses `rep movsb`. At the same time, we have also verified through experiments that the current Rust compiler's built-in memcpy is slower on a page-sized(4KB) memory scale compared to directly using assembly. 

Therefore, we chose to manually write assembly instructions for memory copying in the frame to achieve the same performance as Linux. This ensures fairness during benchmark testing.

As for the two newly added functions in #845 , I believe their current implementation is still incomplete. This is partly because I am not very familiar with assembly code and related CPU features (thanks to @lrh2000  for the explanation about this part in Intel's manual). In fact, a better implementation should use different versions of memory copy based on different CPU architectures and features to achieve optimal performance (I believe Linux takes this approach as well).

Ref: https://elixir.bootlin.com/linux/v6.9.2/source/arch/x86/include/asm/uaccess_64.h#L103

**Comment by junyang-zh** [comment]
> > the current fast_copy_nonoverlapping is a maintenance disaster
> I don't think so. 

@liqinggd IMHO, if this claim is true, we would have to make a copy of every performance sensitive implementations (including `compiler_rt`, `core`, `alloc`, ...) just to have a sense of control over the performance regression. This is not friendly to maintain (and bloats generated code size since you cannot drop your dependency of `compiler_rt`). As for the comparison with Linux, you'll sure have to maintain it if Linux gets updated. Linux chooses to have all those low-level implementations because there's no other choice for it. With Rust we have other choices. And I hardly support that our capabilities are enough to maintain all magical implementations throughout the decades of iteration of Linux and glibc (#623 for example). That could be handled by the `compiler_rt` team.

> Our extensive experience with actual projects has taught us that having the code under our own control is the most reliable approach.

I admit that control is good. But I don't have "extensive experience" with it. So I do not weigh such control over maintainability. Please have me a lesson if applicable.

> but if someone changes the toolchain version or any upstream dependency changes, it can easily introduce performance regressions.

An I have a faith the the upstream is only going better if we don't somehow downgrade dependencies.

In short, I am still not persuaded if #845 is better than #890 in the way to solve the problem.

**Comment by lrh2000** [comment]
@liqinggd Thanks for your explanation.

> In fact, a better implementation should use different versions of memory copy based on different CPU architectures and features to achieve optimal performance. I believe Linux takes this approach as well.

Yes, here is another reference. Different instructions are used depending on whether the CPU has ERMS or FSRM features: https://elixir.bootlin.com/linux/v6.9.2/source/arch/x86/lib/memmove_64.S#L44

>  Since we do not maintain Rust's compiler-rt, [..] Our extensive experience with actual projects has taught us that having the code under our own control is the most reliable approach.

Actually, I agree with your argument, but I disagree with your solution. I think that **the correct conclusion is that we should maintain `compiler_builtins` ourselves** instead of adding a `fast_copy` implementation to aster-frame. Again, my main concern is that all implicit calls to `ptr::copy` like `Vec::clone`, `slice::copy_from_slice` should be fixed together, but adding a `fast_copy` is never possible to fix this problem. It also adds overhead for framework users to remember to use `fast_copy', which is not a good thing either.

Maintaining `compiler_builtins` ourselves does not give much overhead, because most of the `compiler_builtins` are related to floats, but we should not be using floats in the kernel context anyway. So reimplementing utilities in `compiler_builtins` does not give us much overhead, but gives us a significant benefit (as you said, we can choose which version of `memcpy` to use according to CPU capabilities).

But the upstream `compiler-rt` still needs to be fixed, though. That's a separate issue.

**Comment by liqinggd** [comment]
> I think that the correct conclusion is that we should maintain compiler_builtins ourselves.

I agree with that if it does not give much overhead. As you mentioned that other implicit calls may benefit from the `compiler-rt`, it seems that maintaining it would be a better solution.

**Comment by lrh2000** [comment]
> IMHO, if this claim is true, we would have to make a copy of every performance sensitive implementations (including compiler_rt, core, alloc, ...) just to have a sense of control over the performance regression.

Hopefully, few of them have performance related to CPU capabilities.

But if their performance does depend on CPU capabilities, and we want to dynamically control the code at runtime (to choose the best implementation according to CPU capabilities), maintaining them ourselves may be necessary if the upstream does not want to do it (I'm really not sure about the `compiler_builtins` case, though).

There is certainly a tradeoff, and for simplicity I think we can just stick with compile-time features for now (like `target_features = "ermbs"`), and let users choose which features to enable depending on the capabilities of the CPU where they want to run the kernel.

> That could be handled by the compiler_rt team.

It's another option, but I don't think they're going to support runtime detection of CPU capabilities in the near future. But it's certainly possible to try to suggest it.

**Comment by lrh2000** [comment]
> maintaining them ourselves may be necessary if the upstream does not want to do it (I'm really not sure about the `compiler_builtins` case, though).

For the glibc case, it simply sets a number of thresholds at startup (e.g., the `x86_rep_movsb_threshold` tunable). This can be implemented similarly in `compiler_builtins`, but it has non-zero runtime overhead.

For the Linux case, it dynamically rewrites the machine code at boot time (e.g., the `ALTERNATIVE_2` macro). It achieves absolutely zero runtime overhead, but this approach cannot be implemented by `compiler_builtins` because it does not know how to mark a code page as writable (which requires interaction with the kernel).

**Comment by StevenJiang1110** [comment]
Sorry for replying so late.

> No the "Fast Short REP MOVSB" feature is not introduced in our servers.

It seems @junyang-zh and I do not perform experiments on the same server. Our server does have the `fsrm` and `ersm` feature, and the CPU version is Intel(R) Xeon(R) Gold 6342 CPU @ 2.80GHz. 

For the PR #890, I does another experiment on our server again with the flag `RUSTFLAGS="-C target-feature=+ermsb"`. It still shows a spike in the results, while the spike moves left. 

<img width="832" alt="截屏2024-05-29 17 51 12" src="https://github.com/asterinas/asterinas/assets/27764680/ffcc3c2a-e9e2-4ac4-a90f-c41814342195">

```text
size = 3920, nonoverlap = 40.529729ms, ptr::copy = 40.604001ms, fast copy = 37.284707ms
size = 3928, nonoverlap = 41.150899ms, ptr::copy = 40.857337ms, fast copy = 37.36169ms
size = 3936, nonoverlap = 110.937929ms, ptr::copy = 110.942224ms, fast copy = 35.787206ms
size = 3944, nonoverlap = 111.327272ms, ptr::copy = 111.104264ms, fast copy = 37.892802ms
size = 3952, nonoverlap = 40.851041ms, ptr::copy = 41.218022ms, fast copy = 27.812242ms
size = 3960, nonoverlap = 40.873737ms, ptr::copy = 41.12203ms, fast copy = 27.842908ms
size = 3968, nonoverlap = 93.159386ms, ptr::copy = 93.163443ms, fast copy = 34.84191ms
size = 3976, nonoverlap = 93.038694ms, ptr::copy = 93.244204ms, fast copy = 36.829243ms
size = 3984, nonoverlap = 42.167502ms, ptr::copy = 42.343848ms, fast copy = 37.912275ms
size = 3992, nonoverlap = 42.676596ms, ptr::copy = 43.033005ms, fast copy = 37.909282ms
size = 4000, nonoverlap = 98.499437ms, ptr::copy = 98.477202ms, fast copy = 37.915415ms
size = 4008, nonoverlap = 98.47456ms, ptr::copy = 98.525064ms, fast copy = 37.704086ms
size = 4016, nonoverlap = 33.859637ms, ptr::copy = 33.949093ms, fast copy = 27.351831ms
size = 4024, nonoverlap = 33.963182ms, ptr::copy = 33.952357ms, fast copy = 27.410264ms
size = 4032, nonoverlap = 60.397941ms, ptr::copy = 60.683658ms, fast copy = 36.825548ms
size = 4040, nonoverlap = 60.239203ms, ptr::copy = 60.575996ms, fast copy = 37.004838ms
size = 4048, nonoverlap = 32.351118ms, ptr::copy = 31.983343ms, fast copy = 37.737973ms
size = 4056, nonoverlap = 33.326659ms, ptr::copy = 32.860375ms, fast copy = 37.656155ms
size = 4064, nonoverlap = 62.858809ms, ptr::copy = 62.841897ms, fast copy = 36.675405ms
size = 4072, nonoverlap = 62.857607ms, ptr::copy = 62.866928ms, fast copy = 38.217008ms
size = 4080, nonoverlap = 30.585994ms, ptr::copy = 30.47474ms, fast copy = 28.851859ms
size = 4088, nonoverlap = 30.791608ms, ptr::copy = 30.839002ms, fast copy = 28.848413ms
size = 4096, nonoverlap = 42.36443ms, ptr::copy = 42.766529ms, fast copy = 58.268981ms
```
The efficiency of `ptr::copy` is still rather poor when the buffer size is 3936, 3944, 3968, 3974 and so on. But when the buffer size is 4096, the `ptr::copy` seems faster than the assembly version now.

I think the graph in [ruat-lang#comment](https://github.com/rust-lang/rust/issues/125352#issuecomment-2123346254) shoule be the ideal performance results, a smooth line(if the outlier in 4096 does not exist, but this may be a CPU bug). This is reasonable since whatever the underlying instruction is used(either SIMD or `rep mov`, they are both highly-optimized instructions), the CPU should provide the similar performance. 

FYI: [rust-lang: issue](https://github.com/rust-lang/rust/issues/125352)

**Comment by StevenJiang1110** [comment]
Since the results are hardware-related, others may not be able to reproduce the results. I put the experiment results when the buffer size approaches 4096 here.

After redoing the experiment, even without the `"-C target-feature=+ermsb"`, the spike moves left. It is strange for me and conflicts with the original results.

The results with `RUSTFLAGS="-C target-feature=+ermsb"`:
```text
size = 4064, nonoverlap = 62.35492ms, ptr::copy = 60.185434ms, fast copy = 37.660412ms
size = 4072, nonoverlap = 60.204831ms, ptr::copy = 60.047925ms, fast copy = 37.771134ms
size = 4080, nonoverlap = 30.693214ms, ptr::copy = 30.552224ms, fast copy = 25.814345ms
size = 4088, nonoverlap = 30.62865ms, ptr::copy = 30.71643ms, fast copy = 25.83909ms
size = 4096, nonoverlap = 42.590958ms, ptr::copy = 42.86982ms, fast copy = 59.614409ms
size = 4104, nonoverlap = 59.871102ms, ptr::copy = 60.53304ms, fast copy = 60.503608ms
```
The results WITHOUT `RUSTFLAGS="-C target-feature=+ermsb"`:
```text
size = 4064, nonoverlap = 43.726248ms, ptr::copy = 43.659242ms, fast copy = 37.85892ms
size = 4072, nonoverlap = 43.970558ms, ptr::copy = 43.984477ms, fast copy = 37.830849ms
size = 4080, nonoverlap = 30.648833ms, ptr::copy = 30.983275ms, fast copy = 28.734991ms
size = 4088, nonoverlap = 30.547722ms, ptr::copy = 30.681477ms, fast copy = 28.797846ms
size = 4096, nonoverlap = 41.672566ms, ptr::copy = 41.611703ms, fast copy = 62.119747ms
```

**Comment by lrh2000** [comment]
@StevenJiang1110 You need to clarify whether your experiments are in kernel space or not. If you are experimenting in userspace, `ptr::copy` calls your system libc's memcpy, which is a _completely_ different [implementation](https://elixir.bootlin.com/glibc/glibc-2.39/source/sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S#L889) and really has nothing to do with either `compiler_builtins` or `target-feature=+ermsb`.

**Comment by StevenJiang1110** [comment]
> @StevenJiang1110 You need to clarify whether your experiments are in kernel space or not. If you are experimenting in userspace, `ptr::copy` calls your system libc's memcpy, which is a _completely_ different [implementation](https://elixir.bootlin.com/glibc/glibc-2.39/source/sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S#L889) and really has nothing to do with either `compiler_builtins` or `target-feature=+ermsb`.

The experiments are done in userspace. I do not try to do the experiments in kernel space.

**Comment by lrh2000** [comment]
A small update to the following statement:

> I think that **the correct conclusion is that we should maintain `compiler_builtins` ourselves** instead of adding a `fast_copy` implementation to aster-frame.

We don't even have to maintain the whole `compiler_builtins` ourselves. We can always override the default `memcpy`/`memmove` implementations _if we want_, according to [the comments in `compiler_builtins`](https://github.com/rust-lang/compiler-builtins/blob/c04eb9e1afb72bdf943f5e5d77b3812f40526602/src/macros.rs#L28-L33), as quoted below:
```rust
/// All intrinsics functions are marked with #[linkage = "weak"] when
/// `not(windows) and not(target_vendor = "apple")`.
/// `weak` linkage attribute is used so that these functions can be replaced
/// by another implementation at link time. This is particularly useful for mixed
/// Rust/C++ binaries that want to use the C++ intrinsics, otherwise linking against
/// the Rust stdlib will replace those from the compiler-rt library.
```

**Comment by the8472** [comment]
> This is because, during our memory copy performance testing, our comparison target is the Linux kernel, and we have confirmed that it directly uses rep movsb.

Note that, like glibc, the linux kernel has multiple memcpy routines and only uses rep mov on some CPUs.

https://github.com/torvalds/linux/blob/master/arch/x86/lib/copy_page_64.S
https://github.com/torvalds/linux/blob/master/arch/x86/lib/memcpy_64.S

**Comment by workingjubilee** [comment]
The unusual spikes in performance between around 3KiB to 6KiB are due to some of the writes hitting the [memory page](https://en.wikipedia.org/wiki/Page_(computer_memory)) boundary of the machine, as x86-64 machines usually use 4KiB pages.
